src/mlpack/core/optimizers/saga/saga_impl.hpp

```diff
@@ -0,0 +1,175 @@
+#define MLPACK_CORE_OPTIMIZERS_SAGA_SAGA_IMPL_HPP
+
+// In case it hasn't been included yet.
+#include "saga.hpp"
+
+namespace mlpack {
+namespace optimization {
+
+template<typename UpdatePolicyType, typename DecayPolicyType>
+SAGAType<UpdatePolicyType, DecayPolicyType>::SAGAType(
+    const double stepSize,
+    const size_t batchSize,
+    const size_t maxIterations,
+    const double tolerance,
+    const bool shuffle,
+    const UpdatePolicyType& updatePolicy,
+    const DecayPolicyType& decayPolicy,
+    const bool resetPolicy) :
+    stepSize(stepSize),
+    batchSize(batchSize),
+    maxIterations(maxIterations),
+    tolerance(tolerance),
+    shuffle(shuffle),
+    updatePolicy(updatePolicy),
+    decayPolicy(decayPolicy),
+    resetPolicy(resetPolicy)
+{ /* Nothing to do. */ }
+
+//! Optimize the function (minimize).
+template<typename UpdatePolicyType, typename DecayPolicyType>
+template<typename DecomposableFunctionType>
+double SAGAType<UpdatePolicyType, DecayPolicyType>::Optimize(
+    DecomposableFunctionType& function,
+    arma::mat& iterate)
+{
+  // Find the number of functions to use.
+  const size_t numFunctions = function.NumFunctions();
+
+  // To keep track of where we are and how things are going.
+  double overallObjective = 0;
+  double lastObjective = DBL_MAX;
+
+  // Initialize the update policy.
+  if (resetPolicy)
+    updatePolicy.Initialize(iterate.n_rows, iterate.n_cols);
+
+  // Now iterate!
+  arma::mat gradient(iterate.n_rows, iterate.n_cols);
+  arma::mat gradient0(iterate.n_rows, iterate.n_cols);
+  arma::mat iterate0;
+
+  // Find the number of batches.
+  size_t numBatches = numFunctions / batchSize;
+  if (numFunctions % batchSize != 0)
+    ++numBatches; // Capture last few.
+
+  const size_t actualMaxIterations = (maxIterations == 0) ?
+               std::numeric_limits<size_t>::max() : maxIterations;
+
+  // initialize the gradients.
+  size_t effectiveBatchSize = std::min(batchSize, numFunctions);
+  arma::cube tableOfGradients(iterate.n_rows, iterate.n_cols, numBatches);
+  arma::mat avgGradient(iterate.n_rows, iterate.n_cols);
+
+  for (size_t f = 0, b = 0; f < numFunctions;
+    /* incrementing done manually */)
+  {
+    // Find the effective batch size (the last batch may be smaller).
+    effectiveBatchSize = std::min(batchSize, numFunctions - f);
+
+    function.Gradient(iterate, f, tableOfGradients.slice(b),
+                      effectiveBatchSize);
+
+    f += effectiveBatchSize;
+    avgGradient += tableOfGradients.slice(b);
+    b++;
+  }
+  avgGradient /= (double) numBatches; // Calculate average gradient
+
+  for (size_t i = 0; i < actualMaxIterations; ++i)
+  {
+    // Calculate the objective function.
+    overallObjective = 0;
+    for (size_t f = 0; f < numFunctions; f += batchSize)
+    {
+      const size_t effectiveBatchSize = std::min(batchSize, numFunctions - f);
+      overallObjective += function.Evaluate(iterate, f, effectiveBatchSize);
+    }
```

Comment:
If we use `EvaluateWithGradient()` in the loops over batches below, we can cache the objective value of the model (or something very close to it) for way cheaper than this.  Typically looping over all batches to calculate the objective value over all data can be very very time-consuming.  If there is a reason we need the exact calculation here, that's fine with me, but I do want to point out the very large overhead involved.

Reply:
I'm still not clear should we use `EvaluateWithGradient()` every time in place of `Evaluate()` for faster processing or whenever we want objective value and gradient together in an efficient way?

Reply:
`EvaluateWithGradient()` in many cases will provide you with both the objective and the gradient in less time than it takes to call both `Evaluate()` and `Gradient()`.  So if you can restructure the code to only call `EvaluateWithGradient()` instead of `Gradient()` and then cache the objective values it is returning, you don't have to loop over all of the batches and call `Evaluate()`, which can save a significant amount of processing time.

Reply:
Thanks, now its clear to me but I couldn't think of refactoring the code much to use it because here the `gradient` of only a random function is calculated and the random function can repeat itself, so if I try to `Evaluate` and add to the `objectiveValue` then it'll be misleading. The only way I can use it is at the time of initializing gradients so that initial `objectiveValue` will be calculated at the same time using `EvaluateWithGradient()`. It just leads to one less looping over all the functions to calculate objective.

Reply:
The choices are this:

 * We calculate the exact objective value at the end of an epoch and this means calling `Evaluate()` on every point in the dataset.  That might be very large and could take nearly as long as the epoch of training itself.
 * We sum the objectives as we perform the epoch, giving us an estimate of the objective value, not an exact value.

Given that the objective is used specifically for convergence, I think it is fine to go with the latter.  This is what Keras does.  If the parameters are not moving much between two epochs, then it is still true that the objective will converge.

Reply:
Ohh!! I get it now. It does make sense and ultimately what we are after is convergence and not the objective value itself. I'll make the required changes if you could kindly clarify my above points. Thanks.